Modular firearms - MFCS

BlueNexus opened the pull request:
//THIS IS HEAVILY WORK IN PROGRESS//

The MFCS allows players to construct and disassemble firearms using a wide variety of components, allowing the player to customise recoil, accuracy, projectile type, fire rate and mode, projectile speed, loading method and more. All firearms in the game will be replaced with these modular variants. MFCS Firearms may be constructed by R&D or anybody who gets a hold of the parts, however the final part will come in a lock box, to prevent them being made and used for grief. This will hopefully give research something interesting to do, and will add a lot of depth to the firearms in the game. Keep in mind, this project is massively work in progress. Feedback would be appreciated, however.

Current build - 215

BlueNexus committed:
Added the framework files for components

BlueNexus committed:
Added the different chambers

BlueNexus committed:
WIP Modular firearm code

BlueNexus committed:
Changed allowed_projectiles to use caliber

So much cleaner.
Also gave some things a placeholder sprite

BlueNexus committed:
Removed a stray piece of text

BlueNexus committed:
Moved projectile type into chamber/

BlueNexus committed:
Placeholder code for drivers

BlueNexus committed:
Placeholder drivers to work with

BlueNexus committed:
Assembly driver code.

Heavily WIP and untested.

BlueNexus committed:
Adds a debug check

BlueNexus committed:
Duplicate definition fix

BlueNexus committed:
Reworked driver firemodes

BlueNexus committed:
Reworked driver firemodes

BlueNexus committed:
Added loaders

BlueNexus committed:
Moved the C.shell under shell

BlueNexus committed:
Added loaders, and commenting

BlueNexus committed:
Quick patch

BlueNexus committed:
Will test travis

BlueNexus committed:
Revert

BlueNexus committed:
Create energy.dm

BlueNexus committed:
Create automatic

BlueNexus committed:
Create shotgun.dm

BlueNexus committed:
Rename automatic to automatic.dm

BlueNexus committed:
Create sniper.dm

BlueNexus committed:
Create pistol.dm

BlueNexus committed:
Delete energy.dm

BlueNexus committed:
Create beam.dm

BlueNexus committed:
Create special.dm

BlueNexus committed:
Create nuclear.dm

BlueNexus committed:
Create dart.dm

BlueNexus committed:
Added energy loaders

BlueNexus committed:
Var refactor

Will now use a compile_gun() proc to load all the vars, so that premade guns work

BlueNexus committed:
Removed obsolete code

BlueNexus committed:
Wiped core.dm, different purpose

BlueNexus committed:
Continue fix

BlueNexus committed:
Var fix

BlueNexus committed:
Major component refactor

A proc is now used, rather than repeating variables for each time an object is used on the frame. This also means that the proc can be called without attacking the frame, such as if an admin triggers it. This will also be used for dissasembly in the future.

BlueNexus committed:
Further refactor

Super-clean way of doing it, removing dozens of unnecessary lines of code. This is so CLEEEEEEEEEAN

BlueNexus committed:
Removed unnecessary vars

Buildstage has been entirely removed, thanks to the refactor.

BlueNexus committed:
Cleaning up code

BlueNexus committed:
Moved procs into core

BlueNexus committed:
Moved procs into core

BlueNexus committed:
Added testing SRC's

Not sure if these are needed, worth seeing if they fix the compile errors.

BlueNexus committed:
Added barrels

BlueNexus committed:
Added some +fun barrels

Good luck ever getting them though.

BlueNexus committed:
Added barrel handling to core

BlueNexus committed:
Added stocks

BlueNexus committed:
Comments

BlueNexus committed:
Fixed a compile error

BlueNexus committed:
Small fix

BlueNexus committed:
Small fix

BlueNexus committed:
Fixed assemblies eating parts

Hopefully, at least.

BlueNexus committed:
Added assembly handling for sights and stocks

Now for core.

BlueNexus committed:
Added a var for error handling

BlueNexus committed:
Refactored failure messages

BlueNexus committed:
Refactored failure messages and a minor fix

BlueNexus committed:
Glorious refactoring

Took out at least 25 lines. Procs are great.

BlueNexus committed:
Even more refactoring

BlueNexus committed:
Fixed the refactor

BlueNexus committed:
Fixed errors not returning

BlueNexus committed:
Cleaned up code, framework for decon

BlueNexus committed:
Removed unnecessary vars

BlueNexus committed:
Added a list

BlueNexus committed:
Added a part removal proc

BlueNexus committed:
Part removal update

BlueNexus committed:
Hopeful compile fix

BlueNexus committed:
Added attachment handling to disassembly

BlueNexus committed:
Added charge costs

BlueNexus committed:
Create base.dm

BlueNexus committed:
Base projectile file

BlueNexus committed:
Work-in-progress compile proc

BlueNexus committed:
Added casing handling to compile

BlueNexus committed:
Added heat level handling

BlueNexus committed:
Added a heat limit var for later use

BlueNexus committed:
Work-in-progress overheat handling

BlueNexus committed:
More heat handling

BlueNexus committed:
Cooldown handling

Not sure if this will work or not. We'll see.

BlueNexus committed:
Further handling

BlueNexus committed:
Create energy.dm

BlueNexus committed:
Extra quotation fix

BlueNexus committed:
Merge branch 'master' into ModularFirearms

BlueNexus committed:
Comments and punctuation

Very minor, but it's something to do while I figure out fire delays for energy weapons.

MistakeNot4892 commented on code/modules/projectiles/gun.dm:
```diff
@@ -150,6 +150,25 @@
+			return
+		if(src.heat_level = src.heat_cap)
+			user << "<span class='warning'>[src] feels hot in your hands!<span>"
+		if((src.heat_level - src.heat_cap) = 1)
```
Should be elseif or you're running checks that you're already establishing as false earlier in the code.

MistakeNot4892 commented on code/modules/projectiles/gun.dm:
```diff
@@ -179,6 +198,9 @@
 		var/disp = firemode.dispersion[min(i, firemode.dispersion.len)]
 		process_accuracy(projectile, user, target, acc, disp)
+		
+		if(istype (src, /obj/item/weapon/gun/energy))
```
Set a var at L154 and check it here, saves checking the type every single loop iteration.

MistakeNot4892 commented on code/modules/projectiles/guns/modular/assembly.dm:
```diff
@@ -0,0 +1,73 @@
+	var/useSupply = null
+	var/useBullet = null
+
+/obj/item/weapon/modular_firearms/assembly/attackby(obj/item/I as obj, mob/user as mob)
```
Either return after setting part/prereq or make these elseif, you're checking type repeatedly even if it evaluates true in the first block.

MistakeNot4892 commented on code/modules/projectiles/guns/modular/core.dm:
```diff
@@ -0,0 +1,140 @@
+
+/obj/item/weapon/modular_firearms/assembly/proc/process_part(obj/item/I as obj, mob/user as mob) //this should handle processing new parts in the weapon, without relying on the weapon actually being attacked.
+	if(istype(I, /obj/item/weapon/modular_firearms/chassis))
+		if(istype(I, /obj/item/weapon/modular_firearms/chassis/energy))
```
Checking mutually exclusive types repeatedly again.

MistakeNot4892 commented on code/modules/projectiles/guns/modular/core.dm:
```diff
@@ -0,0 +1,140 @@
+		if(chamber.projectile_type) //checking for energy weaponry
+			if(src.isEnergy)
+			else
+				src.msg = "\red A ballistic chamber won't work with an energy chassis!"
```
Colour macros are outdated. Span classes! >:

MistakeNot4892 commented:
Cirra asked me for critique, ftr, this isn't me randomly deciding to talk shit on another codebase's PRs.

A lot of the code here seems bloated or redundant, I feel like you could make good use of associative lists for all the various kinetic/ballistic/etc checks rather than having so much boilerplate for each individual var.

MistakeNot4892 commented on code/modules/projectiles/guns/modular/core.dm:
```diff
@@ -0,0 +1,140 @@
+/obj/item/weapon/modular_firearms/assembly/proc/compile(mob/user as mob)
+	var/type = null
+	if(src.isKinetic)
+		var/new/obj/item/weapon/gun/MFCS/projectile/P as obj
```
'as obj' is deprecated.

BlueNexus committed:
Changed if's to else if's 

Should stop it being checked again and again in the code. Thanks Zuhayr.

BlueNexus committed:
Used a var instead of rechecking istype.

Credit to Zuhayr. Let's see if I did this correctly.

BlueNexus committed:
More handling and a refactor

Credit for lines 119 to 124 to @Zuhayr

BlueNexus committed:
Added some else if's

BlueNexus committed:
Switched to using span classes

BlueNexus committed:
Modifying the P again.

Once again, credit to @Zuhayr. I learned a little bit from this one, so this should help in the future.

BlueNexus committed:
Added barrel compiling

BlueNexus committed:
Added stock compiling

BlueNexus committed:
Added new power supplies

BlueNexus committed:
Update power_cells.dm

BlueNexus committed:
Start of the replacement

Beginning the process of replacing all standard firearms with MFCS firearms.

BlueNexus committed:
Added a heavy assembly

Will handle larger weapons.

BlueNexus committed:
Added mastertype and compiled vars

BlueNexus committed:
Added the framework for recompiling

BlueNexus committed:
Small fix

BlueNexus committed:
Update loader.dm

BlueNexus committed:
Rearrangement

BlueNexus committed:
Set the base parts for projectile weaponry

BlueNexus committed:
Update projectile.dm

BlueNexus committed:
Added the Z8 driver

BlueNexus committed:
Moved automatic weapons over to the MFCS system.

BlueNexus committed:
Moved pistols over to the MFCS system

BlueNexus committed:
Update core.dm

Sprite overrides are now handled, max_shells is no longer overriden for existing weapons, and "mastertypes" have been implemented.

BlueNexus committed:
Compilesprites into gun.dm

Added compilesprite defining to new() and recompile()

BlueNexus committed:
Removed redundant code

BlueNexus committed:
WIP readme
